nilearn/_utils/estimator_checks.py

```diff
@@ -546,6 +557,155 @@ def check_masker_compatibility_mask_image(estimator):
+            ValueError, match="Parameter 'imgs' must be provided to "
+        ):
+            estimator.fit()
+    else:
+        estimator.fit()
+        assert estimator.mask_img_ is None
+
+
+def check_masker_mask_img_from_imgs(estimator):
+    """Check maskers mask_img_ inferred from imgs.
+
+    For (Multi)NiftiMasker and SurfaceMasker:
+    they must have a valid mask_img_ after fit even when no mask was passed.
+    For all other maskers in this case mask_img_ will be None.
+    """
+    if accept_niimg_input(estimator):
+        # Small image with shape=(7, 8, 9) would fail with MultiNiftiMasker
+        # giving mask_img_that mask all the data : do not know why!!!
+        shape = (29, 30, 31)
+        input_img = Nifti1Image(_rng().random(shape), _affine_mni())
+
+    else:
+        input_img = _make_surface_img(2)
+
+    # Except for (Multi)NiftiMasker and SurfaceMasker,
+    # maskers have mask_img_ = None after fitting some input image
+    # when no mask was passed at construction
+    estimator = clone(estimator)
+    assert not hasattr(estimator, "mask_img_")
+
+    estimator.fit(input_img)
+
+    if isinstance(estimator, (NiftiMasker, SurfaceMasker)):
+        if accept_niimg_input(estimator):
+            assert isinstance(estimator.mask_img_, Nifti1Image)
+        else:
+            assert isinstance(estimator.mask_img_, SurfaceImage)
+        load_mask_img(estimator.mask_img_)
+    else:
+        assert estimator.mask_img_ is None
+
+
+def check_masker_mask_img(estimator):
+    """Check maskers mask_img_ post fit is valid.
+
+    If a mask is passed at construction,
+    then mask_img_ should be a valid mask after fit.
+
+    Maskers should be fittable
+    even when passing a non-binary image
+    with multiple samples (4D for volume, 2D for surface) as mask.
+    Resulting mask_img_ should be binary and have a single sample.
+    """
+    if accept_niimg_input(estimator):
+        # Small image with shape=(7, 8, 9) would fail with MultiNiftiMasker
+        # giving mask_img_that mask all the data : do not know why!!!
+        shape = (29, 30, 31)
+
+        mask_data = np.zeros(shape, dtype="int8")
+        mask_data[2:-2, 2:-2, 2:-2] = 1
+        binary_mask_img = Nifti1Image(mask_data, _affine_eye())
+
+        input_img = Nifti1Image(_rng().random(shape), _affine_eye())
+
+        non_binary_mask_img = Nifti1Image(
+            _rng().random((*shape, 2)), _affine_eye()
+        )
+
+    else:
+        binary_mask_img = _make_surface_mask()
+        non_binary_mask_img = _make_surface_img()
+
+        input_img = _make_surface_img(2)
+
+    # happy path
+    estimator = clone(estimator)
+    estimator.mask_img = binary_mask_img
+    assert not hasattr(estimator, "mask_img_")
+
+    estimator.fit()
+
+    if accept_niimg_input(estimator):
+        assert isinstance(estimator.mask_img_, Nifti1Image)
+    else:
+        assert isinstance(estimator.mask_img_, SurfaceImage)
+    load_mask_img(estimator.mask_img_)
```

Comment:
This code repeats a lot, could it be a private function?

Reply:
done